lib/compress/zstd_compress.c

```diff
@@ -1112,19 +1142,30 @@ static void ZSTD_ldm_reduceTable(ldmEntry_t* const table, U32 const size,
 *   rescale all indexes to avoid future overflow (indexes are U32) */
 static void ZSTD_reduceIndex (ZSTD_CCtx* zc, const U32 reducerValue)
 {
-    { U32 const hSize = (U32)1 << zc->appliedParams.cParams.hashLog;
-      ZSTD_reduceTable(zc->hashTable, hSize, reducerValue); }
+    {   U32 const hSize = (U32)1 << zc->appliedParams.cParams.hashLog;
+        ZSTD_reduceTable(zc->hashTable, hSize, reducerValue);
+    }
+
+    if (zc->appliedParams.cParams.strategy != ZSTD_btlazy2) {
+        U32 const chainSize = (U32)1 << zc->appliedParams.cParams.chainLog;
+        ZSTD_reduceTable(zc->chainTable, chainSize, reducerValue);
+    }
 
-    { U32 const chainSize = (zc->appliedParams.cParams.strategy == ZSTD_fast) ? 0 : ((U32)1 << zc->appliedParams.cParams.chainLog);
-      ZSTD_reduceTable(zc->chainTable, chainSize, reducerValue); }
+    if (zc->appliedParams.cParams.strategy != ZSTD_fast) {
+        U32 const chainSize = (U32)1 << zc->appliedParams.cParams.chainLog;
+        if (zc->appliedParams.cParams.strategy != ZSTD_btlazy2)
```

Comment:
Should this be `== ZSTD_btlazy2`?

Reply:
Yes, indeed, it should.
And now I'm wondering why the associated test has not crashed ...

Reply:
OK, both issues must be fixed.
Though indeed, it's not as impactful as I initially feared.

The situation was worse when `ZSTD_DUBT_UNSORTED_MARK` was defined as `0xFFFFFFFF`. In which case, after a reduce, the resulting index was pointing nowhere.

Now, since it's defined as `1`, a reduce operation squash it to `0`, which means "no further candidate". It's a much smaller impact, and I could not create a corruption or crash issue with it. But that could just be due to lack of samples. `ZSTD_preserveUnsortedMark()` just ensures there is no unforeseen side-effect after a reduce.